### DISCUSSION TITLE
Shape inference for stateful modules

### DISCUSSION THREAD

**Discussion by Huizerd** [google/flax#1633] [Q&A] [answered]
Hi all,

I'm working on implementing feedforward/recurrent Flax modules that have an internal state. I'd like to have shape inference for this state, because in the case of e.g. convolutions it can be nontrivial to determine the shape beforehand. Obviously, I also want the code to be as fast as possible. I have three approaches so far:

```python
import functools
import time
from typing import Any, Callable, Iterable, Optional, Tuple

import jax
from jax import jit
import jax.numpy as jnp
import jax.random as random
import flax.linen as nn


PRNGKey = Any
Shape = Iterable[int]
Dtype = Any
Array = Any

InitFn = Callable[[PRNGKey, Shape, Dtype], Array]
ActFn = Callable[[Array], Array]


def constant(scale: float, dtype: Dtype = jnp.float_) -> InitFn:
    def init(key, shape, dtype=dtype):
        dtype = jax.dtypes.canonicalize_dtype(dtype)
        return scale * jnp.ones(shape, dtype)
    return init


class ShapeOut(nn.Module):
    size: int
    w_init: InitFn = nn.initializers.lecun_normal()
    leak_i_init: InitFn = constant(0.9)
    leak_v_init: InitFn = constant(0.9)
    bias_init: InitFn = nn.initializers.ones
    act_fn: ActFn = nn.activation.tanh

    @nn.compact
    def __call__(self, input_: Array, state: Array) -> Tuple[Array, Array]:

        ff = nn.Dense(self.size, use_bias=False, kernel_init=self.w_init)(input_)
        leak_i = self.param("leak_i", self.leak_i_init, (self.size,))
        leak_v = self.param("leak_v", self.leak_v_init, (self.size,))
        bias = self.param("bias", self.bias_init, (self.size,))

        i, v, s = state

        i = i * leak_i + ff
        v = v * leak_v * (1 - s) + i
        s = self.act_fn(v - bias)

        return s, jnp.stack([i, v, s])

    @staticmethod
    def reset_state(state_size: int, shape: Shape) -> Array:
        return jnp.zeros((state_size, *shape))
    

class ShapeInf(nn.Module):
    size: int
    w_init: InitFn = nn.initializers.lecun_normal()
    leak_i_init: InitFn = constant(0.9)
    leak_v_init: InitFn = constant(0.9)
    bias_init: InitFn = nn.initializers.ones
    act_fn: ActFn = nn.activation.tanh

    @nn.compact
    def __call__(self, input_: Array, state: Optional[Array] = None) -> Tuple[Array, Array]:

        ff = nn.Dense(self.size, use_bias=False, kernel_init=self.w_init)(input_)
        leak_i = self.param("leak_i", self.leak_i_init, (self.size,))
        leak_v = self.param("leak_v", self.leak_v_init, (self.size,))
        bias = self.param("bias", self.bias_init, (self.size,))

        state = state if state is not None else self.reset_state(3, ff.shape)
        i, v, s = state

        i = i * leak_i + ff
        v = v * leak_v * (1 - s) + i
        s = self.act_fn(v - bias)

        return s, jnp.stack([i, v, s])

    @staticmethod
    def reset_state(state_size: int, shape: Shape) -> Array:
        return jnp.zeros((state_size, *shape))


class ShapeVar(nn.Module):
    size: int
    w_init: InitFn = nn.initializers.lecun_normal()
    leak_i_init: InitFn = constant(0.9)
    leak_v_init: InitFn = constant(0.9)
    bias_init: InitFn = nn.initializers.ones
    act_fn: ActFn = nn.activation.tanh

    @nn.compact
    def __call__(self, input_: Array) -> Array:

        ff = nn.Dense(self.size, use_bias=False, kernel_init=self.w_init)(input_)
        leak_i = self.param("leak_i", self.leak_i_init, (self.size,))
        leak_v = self.param("leak_v", self.leak_v_init, (self.size,))
        bias = self.param("bias", self.bias_init, (self.size,))

        is_init = self.has_variable("state", "i")
        i = self.variable("state", "i", jnp.zeros, ff.shape)
        v = self.variable("state", "v", jnp.zeros, ff.shape)
        s = self.variable("state", "s", jnp.zeros, ff.shape)

        if is_init:
            i.value = i.value * leak_i + ff
            v.value = v.value * leak_v * (1 - s.value) + i.value
            s.value = self.act_fn(v.value - bias)

        return s.value


@functools.partial(jit, static_argnums=(0,))
def forward(layer, params, input_, state):
    return layer.apply(params, input_, state)


@functools.partial(jit, static_argnums=(0,))
def forward_statevar(layer, params, input_, state):
    return layer.apply({"params": params, **state}, input_, mutable=["state"])


if __name__ == "__main__":
    batch = 2
    input_ = random.uniform(random.PRNGKey(0), (batch, 4))

    hidden = 8

    # outside init needs knowledge of shape; can be difficult with convolutions
    shapeout = ShapeOut(hidden)
    state_shapeout = shapeout.reset_state(3, (batch, hidden))
    params_shapeout = shapeout.init(random.PRNGKey(0), input_, state_shapeout)

    # inside init
    shapeinf = ShapeInf(hidden)
    state_shapeinf = None
    params_shapeinf = shapeinf.init(random.PRNGKey(0), input_, state_shapeinf)

    # state as variable
    shapevar = ShapeVar(hidden)
    state_shapevar, params_shapevar = shapevar.init(random.PRNGKey(0), input_).pop("params")

    time_start = time.time()
    for _ in range(1000):
        output_shape_out, state_shapeout = forward(shapeout, params_shapeout, input_, state_shapeout)
    print(f"ShapeOut: {time.time() - time_start}")

    time_start = time.time()
    for _ in range(1000):
        output_shapeinf, state_shapeinf = forward(shapeinf, params_shapeinf, input_, state_shapeinf)
    print(f"ShapeInf: {time.time() - time_start}")

    time_start = time.time()
    for _ in range(1000):
        output_shapevar, state_shapevar = forward_statevar(shapevar, params_shapevar, input_, state_shapevar)
    print(f"ShapeVar: {time.time() - time_start}")
```

Here, `ShapeOut` has a static method for initializing state outside; `ShapeInf` receives state as an optional argument, allowing shape inference inside; `ShapeVar` saves state as a `variable`, also allowing shape inference inside.

Running the above code gives me:
```
 WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
ShapeOut: 0.07951784133911133
ShapeInf: 0.16239190101623535
ShapeVar: 0.08555960655212402
```

Shape inference with an optional argument is clearly slowest in this case, but it does seem most convenient to me. My question to you is: are there any other ways in which I can have shape inference for the state as well as fast code, or do you know ways in which the above code can be improved? Or maybe there are problems you foresee when using one of these approaches in a training loop? Thanks!

**Comment by marcvanzee** [comment, marked as answer]
I think your numbers are a bit misleading, because you are including the `jit` compiliation in the timings, is this intended? I don't think you will get reliable numbers, and usually you want to either measure how long it take to jit compile, or to execute the compiled functions, but not both.

You can always check when jit complication is triggered this by adding a print statement to your jitted function, like so:

```python
@functools.partial(jit, static_argnums=(0,))
def forward(layer, params, input_, state):
    print('Compiling foward')
    return layer.apply(params, input_, state)
```

Now every time `forward` is compiled it will print this statement. In your case these print statements will be inside the timing loops. In order to avoid this, you should include a "warmup" step before the timings, which will just call the function onces and trigger compliation.

Also, did you know that Colab provides a useful magic function `%timeit` for timing functions automatically?

I've made the above suggested changes to your code and put them in this Colab: https://colab.research.google.com/drive/1Mv2-gvWzBKld8mL15PU1K6MECt8Z-v4s?usp=sharing

The timings are now as folllows on CPU. 
* ShapeOut: 13.6 µs per loop
* ShapeInf: 12.2 µs per loop
* ShapeVar: 18.2 µs per loop

These differences are really tiny! 1µs is 10e−6 seconds. The last one could be a bit slower because you have a few more operations in your variables, but I think it isn't very significant.
